### ISSUE TITLE
An option is taken for the argument of another one

### ISSUE THREAD

**Issue by hgfernan** [jarro2783/cxxopts#100]
Hello, @jarro2783 ! 

1st of all, thanks for sharing an impressive piece of software as cxxopts. 

Please help  me with what's probably my misuse of your library. 

Consider the following sample of code, and further suppose that it's part of an app named tCxxopts 

	options.add_options ()
		("d,debug", "Enable debugging")
		("h,help" , "Help listing")
		("f,file" , "File name", cxxopts::value<std::string>())
		("t,times", "Number of times", cxxopts::value<std::size_t>())
		;

This app only echoes the command line it receives.

I followed the sample code you outlined and also the one by @mmmcell, but for me the following command line
./tCxxopts -f -d

will cause "-d" to become the file name that "-f" expects, while I would prefer that it generated the error 
Option ‘f’ is missing an argument

just as when the command line 
./tCxxopts -f 

is issued

Please tell me what I'm doing wrong here, and what should I do to have cxxopts to behave the way I would like it to. 

All the best,
Hilton

**Comment by hgfernan** [comment]
Dear all, 

just to make clear what I am trying to do, here is a sample program that shows the behavior I'm relating.

```
/**
  *  main.cpp
  * 
  * Created by Melissa Cell on 2/20/17.
  * Copyright © 2017 Melissa Cell. All rights reserved.
  * 
  * Slighly adapted from 
  * <ol>
  *     <li>
  *         <a href="https://github.com/mmmcell/cxxopts_test/blob/master/argparse_git/main.cpp">
  *             mmmcell/cxxopts_test
  *         </a>
  *     </li>
  * </ol>
  * 
 */

#include <cstdlib>  // exit()
#include <iostream>
#include <string>
#include <cxxopts.hpp>

void test (cxxopts::Options& options, int argc, char** argv){    
    auto params = options.parse (argc, argv);
    
    if (params["help"].count ()) {
        std::cout << options.help () << std::endl;
        exit (0);
    }
    
    if (params["debug"].count ()) {
        std::cout << "'-d' or '--debug' was used." << std::endl;
    }
    
    if (params["file"].count ()) {
        std::cout << "'-f' or '--file' was used." << std::endl;
        std::cout << "\tValue used was '" 
                  << params["file"].as<std::string> ()
                  << "'" << std::endl;
    }
    
    if (params["times"].count ()) {
        std::cout << "'-t' or '--times' was used." << std::endl;
        std::cout << "\tValue used was '" 
                  << params["file"].as<std::size_t> ()
                  << "'" << std::endl;
    }
}


int main(int argc, char * argv[]) {
    cxxopts::Options options("Tester", "Tester - A testing program.");
    
    options.add_options ()
        ("d,debug", "Enable debugging")
        ("h,help" , "Help listing")
        ("f,file" , "File name", cxxopts::value<std::string>())
        ("t,times", "Number of times", cxxopts::value<std::size_t>())
        ;
    
    test (options, argc, argv);
    
    return 0;
}

```
When it is compiled by 6.3.0 20170516 (Debian 6.3.0-18+deb9u1) with the command line 
`g++ -W -Wall -o tCxxopts tCxxopts.cpp`

with cxxopts cloned from git, it issues no warnings -- in a seemingly correct compilation. However, when run with the command line 
`./tCxxopts -f -d`

it will issue the result
'-f' or '--file' was used.
        Value used was '-d'

While I agree with overall philosophy of` cxxopts `that seems to be "leave the application logic for the programmer", I think that a good command line parsing library should detect that -d is not a viable value for` -f`, but another option. 

Therefore cxxopts should generate an exception like 
```
terminate called after throwing an instance of 'cxxopts::missing_argument_exception'
  what():  Option ‘f’ is missing an argument
Aborted
```
that raised when one uses the command line 
`./tCxxopts -f`

What do you think ?

All the best, 
Hilton

**Comment by jarro2783** [comment]
Hi, I'm glad you find cxxopts useful. Unfortunately you have not misused it, but that is actually how it was designed.

Rather than looking at whether the next parameter looks like an option (as in starts with a `-`), I just made options that take an argument consume the next option. I deliberately chose to implement it this way for simplicitly, because the logic is already complicated enough with default and implicit arguments as well.

The other argument to make is that `-d` is a perfectly legitimate file name. I know other libraries handle this the way you want, and then you need to work around it with `./-d` if that was ever necessary.

If I was to change this behaviour, it would be quite low on my priority list at the moment. It probably wouldn't be completely trivial either. So for now this is how it works, but I could look at it later.

**Comment by hgfernan** [comment]
Dear @jarro2783, thanks for answering my concerns. 

1st of all, like I said, `cxxopts` is a great library, and an impressive software achievement. Besides that it is fully based in templates, and that's really amazing.  Thanks a lot for sharing your great work 

I'm just like any other user of command line, and so I do hate Linux' `/bin/rm` for not allowing files with names beginning with a dash, the popular hyphen. 

But recently I found a strange behavior in a program I've written. It is related to the behavior I mentioned here, of an option becoming the value of another one. Then I concluded that `rm`behavior is not that bad at all.

`cxxopts` is not alone in that interpretation of command line option. [`gengetopt`](https://www.gnu.org/software/gengetopt/), a library that I used a lot in the past also handles incomplete options the same way as cxxopts. I tried [`getopt`](https://en.wikipedia.org/wiki/Getopt)  and it did behave like `cxxopts` in the case options missing arguments.

But [`argparse`](https://docs.python.org/3/library/argparse.html), the module in Python library that does command line processing, behaves like `rm`, so maybe that's a good starting point for further development. However, with 2385 lines of Python in the [3.5 version](https://github.com/python/cpython/blob/3.5/Lib/argparse.py) of the library, `argparse` is certainly not a simple piece of code. 

All the best,
@hgfernan
